### ISSUE TITLE
Migrate index.js to function component

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#16161]
### Filenames
- [ ] [src/components/Hoverable/index.js](https://github.com/Expensify/App/blob/main/src/components/Hoverable/index.js)
    - **type:** `React.Component`
    - **has state values:** true
    - **has refs:** true
    - **has context:** false
    - **lifecycle methods:** `componentDidMount`,`componentWillUnmount`

### Task
- We currently have some class components in our codebase that we would like to refactor to a function component.
- Here's a link with some general advice on how to refactor a class component to a function component: https://react.dev/reference/react/Component#alternatives
- If you need additional guidance, please ask in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q)
- Test for any regressions and verify that there are no breaking changes

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019c6ea0a458401e24</li>
        <li>Upwork Job ID: 1701785419929559040</li>
        <li>Last Price Increase: 2023-10-04</li>
    </ul>
</details>

**Comment by muxriddinmuqimov77** [comment]
I'd like to work on this issue

**Comment by s-alves10** [comment]
I'd love to work on this.

**Comment by vdem0115** [comment]
I can work on this issue.

**Comment by olexyt** [comment]
I can work on this issue.

**Comment by mkhutornyi** [comment]
I'd like to work on this.

**Comment by aditygrg2** [comment]
Hello, I would like to work on this issue.

**Comment by kacper-mikolajczak** [comment]
Hi, I am Kacper from Callstack - an expert contributor group - and I’d like to work on this issue!

**Comment by blaw023** [comment]
Contributor details
Your Expensify account email: skillgro1@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0167e44d60f727e86a

**Comment by blaw023** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Migrate [src/components/Hoverable/index.js](https://github.com/Expensify/App/blob/main/src/components/Hoverable/index.js) into a functional component. 

### What is the root cause of that problem?

new feature

### What changes do you think we should make in order to solve the problem?

We need to migrate index.js into a functional component. We can do this by using the following hooks: 

- useEffect
- useRef
- useState

**Comment by peterdbarkerUK** [comment]
PR under review

**Comment by tejosiamsyarifudin** [comment]
Contributor details
Your Expensify account email: tejosiam@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01a5bade1e1e9dc027

**Comment by Beamanator** [comment]
Welp, PR got reverted ([here](https://github.com/Expensify/App/pull/28642)) - @kacper-mikolajczak have you had the chance to look into what went wrong?

**Comment by kacper-mikolajczak** [comment]
@Beamanator Not yet, I will investigate it today and get back to you 🔧

**Comment by kacper-mikolajczak** [comment]
Hi @Beamanator, I did not have enough time to investigate it fully so I will continue to work on this tomorrow.

Here is what I found:

It looks like one of the useEffects (line 118) is causing infinite loop when unstable callbacks are provided in form of _onHoverOut_ and _onHoverIn_.

Additionally, in some particular scenarios refs are being lost.

**Comment by kacper-mikolajczak** [comment]
@Beamanator What a journey that is! Today, I once again lost to unknown lands of nested forwarded refs and other creatures. Here is what I've discovered along the way:

- Crashes of **ActionButton** were caused by ref being lost in the `AttachmentPickerWithMenuItems` component at line _240_ where we pass `actionButtonRef` to the children of `Hoverable` (Hoverable itself is not directly rendered there, but it is a child component of a `Tooltip`). When debugging, correct ref seemed to be there but we cannot access it.

- **EmojiPickerButton** has the same idea behind it - it renders `Tooltip` that holds `Hoverable` and then `PressableWithoutFeedback` is passed as a child of `Hoverable` indirectly. In `EmojiPickerButton` we try to retrieve ref of `PressableWithoutFeedback` with no use as it is not present. 

When trying to solve the issue by skipping the `Hoverable` ref handling and using the child's ref directly, I've stumbled upon such error from `BoundsObserver` library:

```
Uncaught Error: Observer should have been installed by the time the component is unmounted
```

It looks like the library modifies the ref as well, which might be seen in its source code:

```javascript
function BoundsObserver(props) {
        var _this = _super.call(this, props) || this;
        _this._childRef = react_1.default.createRef();
        _this._childNode = null;
        _this._observer = null;
        return _this;
    }
    BoundsObserver.prototype.componentDidMount = function () {
        var childRef = this._childRef.current;
        if (!childRef) {
            throw new Error("Reference should have been set by the time the component is mounted");
        }
        var childNode = react_dom_1.default.findDOMNode(childRef);
        if (!(childNode instanceof Element)) {
            throw new Error("Child's corresponding DOM node should be an Element");
        }
        this._childNode = childNode;
        this._observer = this._observe({
            root: childNode,
            activate: this.props.enabled,
        });
    };
    BoundsObserver.prototype.componentDidUpdate = function (prevProps, prevState, snapshot) {
        var _this = this;
        var oldObserver = this._observer;
        if (!oldObserver) {
            throw new Error("Observer should have been installed by the time the component is updated");
        }
        var childNode = this._childNode;
        if (!childNode) {
            throw new Error("Reference should have been set by the time the component is updated");
        }
        /**
         * Handle the potential change of callbacks
         *
         * @returns The new current bounds observer
         */
        var handleCallbackUpdate = function () {
            if (_this.props.onBoundsChange != prevProps.onBoundsChange) {
                oldObserver.disconnect();
                var newObserver = _this._observe({
                    root: childNode,
                    // Ensure that the new observer is active, if the previous one was
                    activate: prevProps.enabled,
                });
                _this._observer = newObserver;
                return newObserver;
            }
            else {
                return oldObserver;
            }
        };
        var currentObserver = handleCallbackUpdate();
        if (this.props.enabled != prevProps.enabled) {
            if (this.props.enabled) {
                this._activate({ observer: currentObserver, root: childNode });
            }
            else {
                this._deactivate({ observer: currentObserver });
            }
        }
    };
    BoundsObserver.prototype.componentWillUnmount = function () {
        var observer = this._observer;
        if (!observer) {
            throw new Error("Observer should have been installed by the time the component is unmounted");
        }
        this._deactivate({ observer: observer });
    };
    BoundsObserver.prototype.render = function () {
        var child = react_1.default.Children.only(this.props.children);
        return react_1.default.cloneElement(child, {
            ref: this._childRef
        });
    };
    ```
I will try to debug it further, and if no meaningful progress will be made, I am going to try to rewrite the component. Also I will contact other devs, so they can help me out. 

Thanks for your patience!

**Comment by Beamanator** [comment]
Thanks so much for your detailed investigation so far! 🙏

**Comment by kacper-mikolajczak** [comment]
Heyo @Beamanator, I managed to fix the issues 🥳 For now, the changes are available in the local [branch](https://github.com/kacper-mikolajczak/ExpensifyApp/tree/refactor/16161/hoverable-class-to-functional).

I am going to create a brand new PR and ping you there when it will be ready. 

Additionally, when I was investigating the component and its logical flow I found out that we might also make the component structured better - I am going to write a proposal for that in some spare time 🤞

**Comment by kacper-mikolajczak** [comment]
@Beamanator, I've linked the new PR that contains all the fixes but is still in draft as I need to test it myself 👍

**Comment by peterdbarkerUK** [comment]
^^ PR is under review

**Comment by peterdbarkerUK** [comment]
Not yet deployed, under QA

**Comment by youssef-lr** [comment]
@kacper-mikolajczak I'm seeing this in production when I resize to mobile view from the console, could it be related to this PR?

<img width="1800" alt="Screenshot 2023-10-19 at 19 06 28" src="https://github.com/Expensify/App/assets/9680864/bee223fc-38fb-4b5d-956a-9d49557ba911">

**Comment by youssef-lr** [comment]
I can confirm that reverting the PR locally fixes it for me

https://github.com/Expensify/App/assets/9680864/5a9206b2-b094-4742-8818-93d967506ed7

**Comment by kacper-mikolajczak** [comment]
Hi @youssef-lr, it looks like the error is caused by the `BoundsObserver` + `Hoverable` combo, which was investigated before and seemed to be fixed. Not for long 😄 Is it critical enough that we want to revert the PR? Wonder if there are more errors, as the original issue is gone.

I will look into it, thanks for the feedback!

**Comment by youssef-lr** [comment]
Nobody has caught this in staging nor prod yet, so I guess it's not easy to reproduce? Weirdly, when I resize the window manually the error doesn't seem to happen. But pressing CMD+Option+J to reveal dev tools which is already set to mobile view triggers it. So I'm not totally sure about how critical it is..we can wait and find out if it's occurring for more users.

**Comment by kacper-mikolajczak** [comment]
@cubuspl42 you might be interested in this 👀 🕵️

**Comment by cubuspl42** [comment]
If there's any problem with [`bounds-observer`](https://github.com/react-ng/bounds-observer), you can create an issue, I might be able to fix it if there's a bug there!

I'm actually surprised nothing turned up for so long, at least nothing I'm aware of.

**Comment by cubuspl42** [comment]
[Known limitations](https://www.npmjs.com/package/@html-ng/bounding-client-rect-observer#known-limitations) are listed here

**Comment by kacper-mikolajczak** [comment]
Thank you for the feedback! I will try to double check what is causing the issue and if the root cause were indeed on the library side, I would create the issue and hopefully resolve it as well 🤞

**Comment by kacper-mikolajczak** [comment]
@youssef-lr  Here is the continuation of an issue described in https://github.com/Expensify/App/pull/29844#issuecomment-1774433140

**Comment by kacper-mikolajczak** [comment]
@youssef-lr I've tried to investigate what is causing the app to crash when responsive view is triggered via devtools. I could not really find a proper reason - it looks like it only happens when we open the devtools for the first time:

- no error when rapidly changing window size
- no error when going between many screen sizes in devtools' responsive view
- no error when app is being opened in devtools's responsive view
- window size does not matter, with larger/smaller screens it's still triggering

It seems that some refs are not being properly attached, or at least not in time for `BoundsObserver` to be ready, thus `BoundsObserver` triggers its error based on missing ref.

I tried to prepare reproducible sample with `BoundsObserver` but it did not give any meaningful results - the app just works.

For today I have no further ideas how to solve the issue. Sorry! I will try to find a way soon!

**Comment by youssef-lr** [comment]
Thank you so much for digging into this @kacper-mikolajczak, appreciate it!

**Comment by kacper-mikolajczak** [comment]
@youssef-lr Here is a solution that should solve the issue here as well: https://github.com/Expensify/App/issues/30278#issuecomment-1781746103

**Comment by peterdbarkerUK** [comment]
Deployed to staging

**Comment by peterdbarkerUK** [comment]
With [this](https://github.com/Expensify/App/pull/30462) now on prod, @youssef-lr can you confirm this is working as expected (and if so, close out the issue)?

**Comment by youssef-lr** [comment]
Working great @peterdbarkerUK!

**Comment by Beamanator** [comment]
Woohoo ! That means we can finally close this out? :D

**Comment by youssef-lr** [comment]
Yup!
